File: Array/linchuanxin/Algorithm/day44_0714_数字转换为十六进制数/day44_leetcode405.java
package day44_0714;



public class day44_leetcode405 {

    public static String toHex(int num) {
//        char[] chars = "0123456789abcdef".toCharArray();
//        StringBuilder sb = new StringBuilder();
//        while (num!=0){
//            sb = sb.append(chars[num%16]);
//            num = num/16;
//        }
//        return sb.toString();
        StringBuffer buffer = new StringBuffer();
        char[] arr ="0123456789abcdef".toCharArray();
        if(num==0) return "0";
        while (num!=0){
            int tmp=num&15;
            buffer.append(arr[tmp]);
            num=num>>>4;
        }

        return buffer.reverse().toString();
    }
    public static void main(String[] args) {
        System.out.println(toHex(26));

        //Integer.toHexString(26);
        System.out.println(26&15);
        System.out.println(10>>>4);
    }

}

File: Array/linchuanxin/Algorithm/day44_0714_数字转换为十六进制数/day44_leetcode405.c
//
// Created by LCX on 2020/7/14.
//
#include <stdio.h>
#include <malloc.h>
#include <string.h>

char *reverse(char *str)
{
    if(str==NULL)
        return NULL;
    int len=strlen(str);
    int i=0;
    while(i<len-i-1){
        str[i]=str[i]^str[len-i-1];
        str[len-i-1]=str[i]^str[len-i-1];
        str[i]=str[i]^str[len-i-1];
        i++;
    };
    return str;

}



char * toHex(int num){
    char arr[]="0123456789abcdef";
    int count = -1;
    char *str = (char*)malloc(sizeof(char)*100);
    unsigned int num1 = num;
    while (num1 != 0 )
    {
        int index = num1&15;
        str[++count] =  arr[index];
        num1 = num1 >> 4;
    }

    if(str==NULL)
        return NULL;
    int len=strlen(str);
    int i=0;
    while(i<len-i-1){
        str[i]=str[i]^str[len-i-1];
        str[len-i-1]=str[i]^str[len-i-1];
        str[i]=str[i]^str[len-i-1];
        i++;
    };
    return str;
}





int main(void)
{
    printf("%s\n",toHex(26));
   // printf("%s\n",toHex(26));
    //char arr[]="0123456789abcdef";
//    char * res = (char*)malloc(9);
//    res = arr[10];
   // printf("%s\n",arr);
}